Initial input for the creation of formal specification documents (#72)
International-Data-Spaces-Association/IDS-G: Specifications/model/information.model.md

# Dataspace Information Model

The information model defines the core concepts, entities, and relationships that underpin a `Dataspace`.

## 1 Terminology

### Dataspace

> A `Dataspace` is a set of technical services that facilitate interoperable asset sharing between entities.

Author: Can you propose an alternative formulation?

> A `Dataspace` is a set of technical services that facilitate interoperable asset sharing between entities.

Reviewer: Maybe something like:

"A `dataspace` is a defined set of entities that exchange data under contractually agreed terms of use. Data in this context does not necessarily mean a dataset or a file, but a determinable set of information (also in the form of an application or accesses to services or similar). From a business point of view, a `dataspace` is an association of a set of participants who want to exchange data offers. From a technical point of view, a `dataspace` is a set of technical services that facilitate interoperable asset sharing between entities.

> A `Dataspace` is a set of technical services that facilitate interoperable asset sharing between entities.

Reviewer: Maybe something like:

"A `dataspace` is build of a defined set of entities that exchange data under contractually agreed terms of use. Data in this context does not necessarily mean a dataset or a file, but a determinable set of information (also in the form of an application or accesses to services or similar). From a business point of view, a `dataspace` is an association of a set of participants who want to exchange data offers. From a technical point of view, a `dataspace` is a set of technical services that facilitate interoperable asset sharing between entities.

> A `Dataspace` is a set of technical services that facilitate interoperable asset sharing between entities.

Author: IMO we should only focus on the technical definition in these documents because they are specifications. The other parts of your definition would fit well in the RAM documents and supporting material.

> ### DataspaceAuthority

Author: +1, let's see what others think

A `DataspaceAuthority` is an entity that manages a `Dataspace`.

### Participant

A `Participant` is a `Dataspace` member that provides and/or consumes assets.

### ParticipantAgent

A `ParticipantAgent` is a technology system that performs operations on behalf of a `Participant`.

### IdentityProvider

An `IdentityProvider` is a trusted technology system that creates, maintains, and manages identity information for a `Participant` and `ParticipantAgents`.

### CredentialIssuer

A `CredentialIssuer` is a trusted technology system that issues verifiable credentials for a `Participant` and `ParticipantAgents`.

### ClearingHouse (TBD)

A `ClearingHouse` is a trusted technology system that records and verifies domain events.

### DataspaceRegistrationService

> A `DataspaceRegistrationService` is a technology system that maintains the state of `Participants` in a `Dataspace`.

Reviewer 2: We used to discuss this and have this mental model of a Support Organization where an entity maintains the state of a participant and then updates the status in DAPS/ParIS/CA *somehow*. It was clear in the discussion that any kind of service is required to do so. We did not define this, yet. Now, talking about decentralized models, we should rethink this now and update this mental model.

> A `DataspaceRegistrationService` is a technology system that maintains the state of `Participants` in a `Dataspace`.

Reviewer 2: How do we build this on the Trust Anchor for our system?

> A `DataspaceRegistrationService` is a technology system that maintains the state of `Participants` in a `Dataspace`.

Reviewer 3: Let's open a ticket for it to enable the discussion.

> A `DataspaceRegistrationService` is a technology system that maintains the state of `Participants` in a `Dataspace`.

Reviewer 2: > Let's open a ticket for it to enable the discussion.

see #76

> A `DataspaceRegistrationService` is a technology system that maintains the state of `Participants` in a `Dataspace`.

Reviewer: ```suggestion
A `Dataspace Registration Service` is a technology system that maintains the state of `Participants` in a `Dataspace`.
```

> ### Asset

Reviewer 3: Let's move from Resource/Representation/Artefact to dcat:DataSet or similar term.

Data or a technical service that can be shared by a `Participant`.

### Policy

A set of rules, duties, and obligations that define the terms of use for an `Asset`.

### Offer

A concrete `Policy` associated with a specific `Asset`.

### Agreement

A concrete `Policy` associated with a specific `Asset` that has been signed by both the provider and consumer `Participants`.

### Catalog

A collection of entries representing `Assets` and their `Offers` that is advertised by a provider `Participant`.

> ### CatalogService

Reviewer 3: From 13.10.22: Separate component discovery functionality from dataset discovery functionality.

> ### CatalogService

Reviewer: ```suggestion
### Catalog Service
```

> A `PariticipantAgent` that makes a `Catalog` accessible to `Participants`.

Reviewer: ```suggestion
A `Pariticipant Agent` that makes a `Catalog` accessible to `Participants`.
```

> ### Connector (DataService)

Reviewer 3: I like the way of basing the Connector definition on the ParticipantAgent rather than saying it is a [server](https://github.com/International-Data-Spaces-Association/IDS-G/tree/main/Glossary#connector). Let's enhance it a bit further to make clear whether a connector may appear as a service, as a deployed software asset, as a physical server...

Author: It could be any of those, but they are implementation details. I was using "technology system" to avoid over-specifying what it is.

> ### Connector (DataService)

Reviewer: ```suggestion
### Connector (Data Service)
```

> A `PariticipantAgent` that produces `Agreements` and manages `Asset` sharing.

Reviewer: ```suggestion
A `Pariticipant Agent` that produces `Agreements` and manages `Asset` sharing.
```

### Contract Negotiation

A set of interactions between a provider `Connector` and consumer `Connector` that establish an `Agreement`.

> ### Asset Transfer

Reviewer: ```suggestion
### Transfer Process
```
To align with the process naming, just call this transfer process?

- A `Participant` is a member of one or more `Dataspaces`. A participant registers `Participant Agents` that perform tasks on its behalf.
- A `Participant Agent` performs tasks such as publishing a catalog or engaging in an asset transfer. In order to accomplish these tasks, a participant agent may
  use a _**verifiable presentation**_ generated from a _**credential**_ obtained from a third-party issuer. A participant agent may also use an _**ID token**_ issued by a
  third-party identity provider. Note that a participant agent is a logical construct and does not necessarily correspond to a single runtime process.
- An `Identity Provider` is a trust anchor that generates `ID tokens` used to verify the identity of a `Participant Agent`. Multiple identity providers may operate in
  a dataspace. The identity standard used by a provider is not defined but could be _OAuth2_ or _Decentralized Identifiers using did:web_. An identity provider may be a third-party
  or a participant itself (for example, in the case of decentralized identifiers).
- A `Credential Issuer` issues _verifiable credentials_ used by participant agents to allow access to assets and verify usage control.

> The diagram below depicts the relationships between `ParticipantAgent` types:

Reviewer: ```suggestion
The diagram below depicts the relationships between `Participant Agent` types:
```

![](./im.participant.entities.png)

> - A `CatalogService` is a `Participant Agent` that makes a [DCAT Catalog](https://www.w3.org/TR/vocab-dcat-3/#Class:Catalog) available to other participants.

Reviewer: ```suggestion
- A `Catalog Service` is a `Participant Agent` that makes a [DCAT Catalog](https://www.w3.org/TR/vocab-dcat-3/#Class:Catalog) available to other participants.
```

- A `Catalog` contains one or more `Asset Entries`, which are [DCAT Datasets](https://www.w3.org/TR/vocab-dcat-3/#Class:Dataset). A `Catalog` also contains **_at least one_**
  [DCAT DataService](https://www.w3.org/TR/vocab-dcat-3/#Class:Data_Service) that references a `Connector` where assets may be obtained.

> - An `Asset Entry` has **one** `Offer`, which is an [ODRL Offer](https://www.w3.org/TR/odrl-model/#policy-offer) describing the _usage control policy_ associated with the asset.

Reviewer 3: We also need a mechanism to describe access policies on asset entries, something which is missing for a long time...

Reviewer: I don't think this has to be part of these documents here as access policies on assets are not communicated via the protocol. However, I agree, this has to be mentioned in some connector architecture/functionality documents.

Reviewer: I don't think this has to be part of these documents here as access policies on assets are not communicated via the protocol. However, I agree, this has to be mentioned in some connector architecture/functionality documents.

> - A `Connector` is a `Participant Agent` that performs `Contract Negotiation` and `Asset Transfer` operations with another connector. An outcome of a `ContractNegotiation` may

Reviewer 3: Does this imply that a data plane implementation is always an integral part of a connector?

Reviewer: Maybe better call it "A `Connector` is a `Participant Agent` that performs `Contract Negotiation` and orchestrates `Asset Transfer` operations with another connector."?

and `Participant Agent` have no representation in the protocol message flows that constitute dataspace interactions. This section outlines the classes that comprise the concrete
elements of the information model, i.e. those that are represented in protocol message flows.

### 2.2.1 Catalog

A `Catalog` is a [DCAT Catalog](https://www.w3.org/TR/vocab-dcat-3/#Class:Catalog) with the following attributes:

- 0..N  `Asset Entries`. Since a catalog may be dynamically generated for a request based on the requesting participant's credentials it is possible for it to contain 0 matching
  asset entries.  (DCAT PROFILE)
- 1.N [DCAT DataService](https://www.w3.org/TR/vocab-dcat-3/#Class:Data_Service) that references a `Connector` where assets may be obtained.  (DCAT PROFILE)

### 2.2.1 Asset Entry

An `Asset Entry` is a [DCAT Dataset](https://www.w3.org/TR/vocab-dcat-3/#Class:Dataset) with the following attributes:

- 1..N `hasPolicy` attributes that contain an ODRL `Offer` defining the usage control policy associated with the asset. **_Offers must NOT contain any target attributes. The

>   target of an offer is the asset associated with the containing asset entry._** (ODRL PROFILE)

Reviewer 3: This means that an Offer can only have one asset but there is a need to maintain one offer for a set of (small/cheap) data assets.

Reviewer: You could either negotiate the usage of each asset in a separate negotiation process. Or you define a set of assets as one asset, to again have the 1..1 relation.

Reviewer 3: Why not allowing more than one entry? Either a set with explicit mentions of the data asset ids or a pattern (all data assets which ids follow the pattern 'data_asset_[0-9]')?

Author: We should be careful with definitions and their relations. As @juliapampus alluded to, an asset (dataset) is the atomic unit of data sharing. An offer is only associated with one asset. An asset may be associated (hasPolicy) with a multiplicity of offers. By extension, a negotiation involves **exactly one** asset. That said, an asset may consist of a single data file, a collection of files, or a non-finite data stream. To keep things manageable from both a specification and implementation perspective, asset "atomicity" must be a conceptual bedrock. Beyond this, the specifications do not need to model the contents of an asset.      

Solving the (important) use case where someone wants to have an easy way to define a policy that is "applied" to a group of assets is an implementation detail that I do not think should be defined in the specification as it is a management feature. 

In the EDC, we enabled this use case by introducing the concept of a [ContractDefinition](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/a418469c8929403613bce132e8fa55678e24a2cf/spi/control-plane/contract-spi/src/main/java/org/eclipse/dataspaceconnector/spi/types/domain/contract/offer/ContractDefinition.java). A `ContractDefinition` is associated with an access policy, a contract policy, and an asset selector expression. The latter is used to resolve a set of assets the definition applies to (basically a query). `Offers` are generated from that combination on the fly.

Other implementations may choose a different approach. We must associate an `Offer` with exactly one asset. How an `Offer` is created is an implementation detail we don't need to address in the specification.

- 1..N [DCAT Distributions](https://www.w3.org/TR/vocab-dcat-3/#Class:Distribution). Each distribution must have at least one `DataService` which specifies where the distribution
  is obtained. Specifically, a `DataService` specifies the endpoint for initiating a `ContractNegotiation` and `AssetTransfer`. (DCAT PROFILE)

### 2.2.1 Offer

An `Offer` is an [ODRL Offer](https://www.w3.org/TR/odrl-model/#policy-offer) with the following attributes:

> - An ODRL `uid` that is a unique UUID. (ODRL PROFILE)

Reviewer 3: I have never understood why ODRL demands the `uid` property instead of simply using the subject id itself (`@id` in JSON-LD)...

Author: In fairness, it may be the result of ODRL predating JSON-LD and JSON itself

- The `Offer` must be unique to a dataset since the target of the offer is derived from its enclosing context.

> - The `Offer` must NOT include an explicit `target` attribute.

Reviewer: ```suggestion
- The `Offer` must NOT include an explicit `target` attribute.
- An `Offer` is to be handled immutable. If an `Offer` needs to be corrected, a new `Offer` should be added to the negotiation process, or the process is cancelled and a new one is started.
```
Maybe add the same hint as for the agreement (suggestion below).

### 2.2.1 Agreement

An `Agreement` is an [ODRL Agreement](https://www.w3.org/TR/odrl-model/#policy-agreement) with the following attributes:

- The `Agreement` class must include one `target` attribute that is the UUID of the asset the agreement is associated with. An agreement is therefore associated with **EXACTLY

>   ONE** asset. (ODRL PROFILE)

Reviewer 3: An agreement must not be changed after its creation but must be recreated with a new identifier if needed.

>   ONE** asset. (ODRL PROFILE)

Reviewer: ```suggestion
  ONE** asset. (ODRL PROFILE)
- An  `Agreement` MUST NOT be changed after its creation but MUST be recreated with a new identifier if needed. This is to be done by initiating a new negotiation process.
```
Picking up the comment from @sebbader.

Author: I think if we use this text or something similar, it should be in the contract negotiation protocol spec, not here.
